From 70ebe3c826d4ed8a4466d4dfe4e7122ea861b843 Mon Sep 17 00:00:00 2001
From: Grzegorz Prajsner <grzegorz.prajsner@arm.com>
Date: Fri, 19 Jun 2015 13:19:47 +0100
Subject: [PATCH 18/20] drm/exynos: switch to using standard fb ioctl

Switch to using standardized IOCTL instead of using exynos
specific one for getting the DMA buffer pointer.  This is achieved
by implementing this function:

  fb_ops.fb_dmabuf_export

Signed-off-by: Guillaume Tucker <guillaume.tucker@arm.com>
---
 drivers/gpu/drm/exynos/exynos_drm_fbdev.c | 57 +++++++++----------------------
 1 file changed, 17 insertions(+), 40 deletions(-)

diff --git a/drivers/gpu/drm/exynos/exynos_drm_fbdev.c b/drivers/gpu/drm/exynos/exynos_drm_fbdev.c
index 470d435..c6eb664 100644
--- a/drivers/gpu/drm/exynos/exynos_drm_fbdev.c
+++ b/drivers/gpu/drm/exynos/exynos_drm_fbdev.c
@@ -18,13 +18,14 @@
 #include <drm/drm_crtc_helper.h>
 #include <drm/exynos_drm.h>
 
+#include <linux/dma-buf.h>
+
 #include "exynos_drm_drv.h"
 #include "exynos_drm_fb.h"
 #include "exynos_drm_fbdev.h"
 #include "exynos_drm_gem.h"
 #include "exynos_drm_iommu.h"
-#define IOCTL_GET_FB_DMA_BUF _IOWR('m',0xF9, __u32 )
-#include <linux/dma-buf.h>
+
 #define NUM_BUFFERS 3
 
 #define MAX_CONNECTOR		4
@@ -65,60 +66,36 @@ static int exynos_drm_fb_mmap(struct fb_info *info,
 	return 0;
 }
 
-static u32 exynos_fb_get_dma_buf(struct fb_info *info)
+static struct dma_buf *exynos_fb_get_dma_buf(struct fb_info *info)
 {
-	int fd = -1;
+	struct dma_buf *buf = NULL;
 	struct drm_fb_helper *helper = info->par;
 	struct drm_device *dev = helper->dev;
 	struct exynos_drm_fbdev *exynos_fbd = to_exynos_fbdev(helper);
 	struct exynos_drm_gem_obj *exynos_gem_obj = exynos_fbd->exynos_gem_obj;
 
 	if( dev->driver->gem_prime_export ) {
-		struct dma_buf *buf = NULL;
 		buf = dev->driver->gem_prime_export( dev, &exynos_gem_obj->base, O_RDWR);
 		if(buf) {
-			fd = dma_buf_fd(buf, O_RDWR);
 			drm_gem_object_reference(&exynos_gem_obj->base);
 		}
 	}
 
-	return fd;
-}
-
-static int fb_ioctl(struct fb_info *info, unsigned int cmd,
-            unsigned long arg)
-{
-	int ret;
-
-	switch (cmd) {
-	case IOCTL_GET_FB_DMA_BUF:
-		u32 __user *out_ptr = (u32 __user *)arg;
-		u32 buf_fd = exynos_fb_get_dma_buf(info);
-		if(buf_fd == -1) {
-			ret = -ENOMEM;
-			break;
-		}
-		ret = put_user(buf_fd, out_ptr);
-		break;
-	default:
-		ret = -ENOTTY;
-	}
-
-	return ret;
+	return buf;
 }
 
 static struct fb_ops exynos_drm_fb_ops = {
-	.owner		= THIS_MODULE,
-	.fb_mmap        = exynos_drm_fb_mmap,
-	.fb_fillrect	= cfb_fillrect,
-	.fb_copyarea	= cfb_copyarea,
-	.fb_imageblit	= cfb_imageblit,
-	.fb_check_var	= drm_fb_helper_check_var,
-	.fb_set_par	= drm_fb_helper_set_par,
-	.fb_blank	= drm_fb_helper_blank,
-	.fb_pan_display	= drm_fb_helper_pan_display,
-	.fb_setcmap	= drm_fb_helper_setcmap,
-	.fb_ioctl = fb_ioctl,
+	.owner			= THIS_MODULE,
+	.fb_mmap		= exynos_drm_fb_mmap,
+	.fb_fillrect		= cfb_fillrect,
+	.fb_copyarea		= cfb_copyarea,
+	.fb_imageblit		= cfb_imageblit,
+	.fb_check_var		= drm_fb_helper_check_var,
+	.fb_set_par		= drm_fb_helper_set_par,
+	.fb_blank		= drm_fb_helper_blank,
+	.fb_pan_display		= drm_fb_helper_pan_display,
+	.fb_setcmap		= drm_fb_helper_setcmap,
+	.fb_dmabuf_export	= exynos_fb_get_dma_buf,
 };
 
 static int exynos_drm_fbdev_update(struct drm_fb_helper *helper,
-- 
1.9.1

